File: api/categories.py
from api._lib.blob_io import load_df
from api._lib.core.category_extraction import extract_categories, preferred_category_column
from api._lib.core.feed_processor import apply_column_selection, apply_filters
from api._lib.handler_utils import BaseHandler, PREVIEW_ROWS


class handler(BaseHandler):
    def do_POST(self):
        try:
            body = self._json_body()
            raw_df = load_df(body["rawDfUrl"])
            kept = apply_column_selection(raw_df, body.get("keepMap", {}))
            filtered = apply_filters(kept, body.get("filters", {}))

            default_src = preferred_category_column(filtered)
            cat_src_col = body.get("catSrcCol") if body.get("catSrcCol") in filtered.columns else default_src

            preview_df = extract_categories(
                filtered,
                cat_src_col,
                bool(body.get("wantMain")),
                bool(body.get("wantPenultimate")),
                bool(body.get("wantFinal")),
            )
            extracted_cols = [
                c for c in ["Main Category", "Penultimate Category", "Final Category"]
                if c in preview_df.columns
            ]
            preview_cols = [cat_src_col] + extracted_cols
            dedup = (
                preview_df[preview_cols]
                .assign(_count=1)
                .groupby(preview_cols, as_index=False)
                .agg(_count=("_count", "sum"))
                .rename(columns={"_count": "Product Count"})
                .sort_values("Product Count", ascending=False)
            )

            self._send_json(200, {
                "preview": dedup.head(PREVIEW_ROWS).fillna("").astype(object).to_dict(orient="records"),
                "allCols": list(filtered.columns),
                "catSrcCol": cat_src_col,
            })
        except Exception as e:
            self._send_json(500, {"error": str(e)})

File: api/categories.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'
import { sendError } from '../api/_lib/handlerUtils'
import { applyColumnSelection, applyFilters } from '../api/_lib/core/feedProcessor'
import { extractCategories, preferredCategoryColumn } from '../api/_lib/core/categoryExtraction'
import { loadRows } from '../api/_lib/sessionStore'

export default async function handler(req: VercelRequest, res: VercelResponse) {
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' })

  try {
    const { rawDfUrl, keepMap, filters = {}, catSrcCol, wantMain, wantPenultimate, wantFinal } = req.body as {
      rawDfUrl: string
      keepMap?: Record<string, boolean>
      filters?: Record<string, string[]>
      catSrcCol?: string
      wantMain?: boolean
      wantPenultimate?: boolean
      wantFinal?: boolean
    }

    let rows = await loadRows(rawDfUrl)
    if (keepMap) rows = applyColumnSelection(rows, keepMap)
    if (Object.keys(filters).length) rows = applyFilters(rows, filters)

    const cols = Object.keys(rows[0] ?? {})
    const srcCol = catSrcCol ?? preferredCategoryColumn(cols) ?? cols[0]

    const w = { wantMain: wantMain ?? false, wantPenultimate: wantPenultimate ?? false, wantFinal: wantFinal ?? false }
    const selectedCols: string[] = [
      ...(w.wantMain        ? ['Main Category']        : []),
      ...(w.wantPenultimate ? ['Penultimate Category'] : []),
      ...(w.wantFinal       ? ['Final Category']       : []),
    ]

    let preview: Record<string, string>[] = []

    if (selectedCols.length) {
      // One or more segments selected — show source value + extracted columns,
      // deduplicated by source path and sorted alphabetically.
      const extracted = extractCategories(rows, srcCol, w.wantMain, w.wantPenultimate, w.wantFinal)
      const seen = new Set<string>()
      for (const row of extracted) {
        const srcVal = String(row[srcCol] ?? '')
        if (seen.has(srcVal)) continue
        seen.add(srcVal)
        const entry: Record<string, string> = { [srcCol]: srcVal }
        for (const col of selectedCols) entry[col] = String(row[col] ?? '')
        preview.push(entry)
      }
      preview.sort((a, b) => String(a[srcCol]).localeCompare(String(b[srcCol])))
    } else {
      // No segments selected — show unique source values with product counts so
      // the user can see what's in the column before deciding what to extract.
      const counts = new Map<string, number>()
      for (const row of rows) {
        const v = String(row[srcCol] ?? '')
        counts.set(v, (counts.get(v) ?? 0) + 1)
      }
      preview = [...counts.entries()]
        .sort((a, b) => b[1] - a[1])
        .map(([value, count]) => ({ [srcCol]: value, 'Product Count': String(count) }))
    }

    // Cap at 500 rows to keep the table manageable
    if (preview.length > 500) preview = preview.slice(0, 500)

    res.status(200).json({
      preview,
      allCols: cols,
      catSrcCol: srcCol,
    })
  } catch (err) {
    sendError(res, err)
  }
}
